pkg/collector/component.go

```diff
@@ -0,0 +1,277 @@
+package collector
+
+import (
+	"context"
+	"io"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/rs/cors"
+	"github.com/spf13/viper"
+	"github.com/uber/jaeger-lib/metrics"
+	"github.com/uber/tchannel-go"
+	"github.com/uber/tchannel-go/thrift"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/jaegertracing/jaeger/cmd/collector/app"
+	"github.com/jaegertracing/jaeger/cmd/collector/app/builder"
+	"github.com/jaegertracing/jaeger/cmd/collector/app/grpcserver"
+	"github.com/jaegertracing/jaeger/cmd/collector/app/sampling"
+	"github.com/jaegertracing/jaeger/cmd/collector/app/sampling/strategystore"
+	"github.com/jaegertracing/jaeger/cmd/collector/app/zipkin"
+	clientcfgHandler "github.com/jaegertracing/jaeger/pkg/clientcfg/clientcfghttp"
+	"github.com/jaegertracing/jaeger/pkg/healthcheck"
+	"github.com/jaegertracing/jaeger/pkg/recoveryhandler"
+	ss "github.com/jaegertracing/jaeger/plugin/sampling/strategystore"
+	"github.com/jaegertracing/jaeger/plugin/storage"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+	jc "github.com/jaegertracing/jaeger/thrift-gen/jaeger"
+	sc "github.com/jaegertracing/jaeger/thrift-gen/sampling"
+	zc "github.com/jaegertracing/jaeger/thrift-gen/zipkincore"
+)
+
+var _ (io.Closer) = (*Component)(nil)
+
+// Component returns the collector as a manageable unit of work
+type Component struct {
+	// required to start a new collector
+	serviceName          string
+	logger               *zap.Logger
+	vp                   *viper.Viper
```

Comment:
I think this should not be here. It's better to follow the already established pattern:
  * instantiate component
  * register flags with Viper
  * component.InitFromViper()
  * component.Start()

Reply:
Yeah, I need to clean this up a bit, as I just wanted to do a quick PoC and see check the interest in it. Same applies to most of the other comments in this PR at this state, which is why I'm marking them as "resolved".